### DISCUSSION TITLE
`onInitialize` event doesn't trigger for LSP server

### DISCUSSION THREAD

**Discussion by TracerDS** [microsoft/vscode-discussions#1059] [Extension Development QnA]
Hi. Im writing a LSP Server for an app (autocompletion, linting, etc.).
Im a total beginner and dont really know anything about LSPs apart from the vague info from docs.
But I have encountered an issue: `onInitialize` doesn't seem to trigger.
The docs also dont say anything about the `vscode-languageclient` and `vscode-languageserver` packages so the only thing I can do is to look up existing extensions and edit my LSP based on their code which is not efficient and may be filled with bugs
```ts
// server.ts
import * as lspserver from 'vscode-languageserver/node'

const LSPOptions = lspserver.ProposedFeatures.all
const LSPConnection = lspserver.createConnection(LSPOptions)

LSPConnection.onInitialize((params: lspserver.InitializeParams) => {
    LSPConnection.sendNotification('testnotify', 'initializing...')
})
LSPConnection.onInitialized((params: lspserver.InitializedParams) => {
    LSPConnection.sendNotification('testnotify', 'Initialized!')
})
```
```ts
// client.ts
import * as vscode from 'vscode'
import * as lspclient from 'vscode-languageclient/node'

let client: lspclient.LanguageClient

export function activate(context: vscode.ExtensionContext) {
    // ...
    client = new lspclient.LanguageClient(packageJSON.name, packageJSON.displayName,
        serverOptions, clientOptions
    )
    client.onNotification("testnotify", (value) => {
        console.log(value)
    })
    // ...
}
```
Output:
```
"Initialized!"
```

**Comment by theReynald** [comment]
Hi TracerDS.

Sounds tricky but here are a few things you can check:

1. Ensure Proper Setup: Make sure that your LSP server is properly configured and initialized before attempting to handle requests. This includes setting up the server capabilities correctly and establishing the connection with the client.

2. Check Client Initialization: Ensure that your client is properly initialized and connected to the server. It seems like your client is receiving the Initialized notification, which indicates that the server has successfully initialized, but it's possible that there might be a delay or some other issue preventing the onInitialize event from firing as expected.

3. Debugging Server: Add some debugging statements or logging to your server code to see if the onInitialize event is being triggered at all. This will help you determine if the issue lies with the event not firing or with the notification not being sent correctly.

4. Review Documentation: While the official documentation may not cover all aspects of using the vscode-languageclient and vscode-languageserver packages, it's still a good idea to review any available documentation or examples provided by the package maintainers. You may find additional information or insights that could help resolve your issue.

5. Check for Errors: Look for any errors or warnings in your server and client code, as well as in the output or console logs. Errors or exceptions could indicate underlying issues that need to be addressed.

6. Test with Minimal Setup: Simplify your server and client code to the bare minimum required to reproduce the issue. This will help isolate the problem and make it easier to identify the cause.

Hopefully by following these steps and carefully reviewing your code, you should be able to identify and resolve the issue with the onInitialize event not triggering as expected in your LSP server.

/r

**Reply by TracerDS** [reply]
Docs dont cover LSP stuff at all. Only vague info and redirections to basic LSP examples. In fact even in [basic LSP example](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-sample) the `onInitialize` event doesn't fire so its not really an issue on my part (unless the Node and TypeScript are misconfigured somehow)

**Comment by TracerDS** [comment]
any help?

**Comment by imbant** [comment]
For me, in the language server, I send request **after** 3000ms in `onInitialized` callback. And in the language client, I register request callback **after** 500ms in `onReady` callback. It works fine for me.
My code is like this:

Server:
```ts
LSPConnection.onInitialized(async () => {
    setTimeout(async () => {
        const result = await connection.sendRequest(
            // ...
        );
    }, 3000);
});
```

Client:
```ts
      // Start the client. This will also launch the server
      client.start();

      client.registerProposedFeatures();

      client.onReady().then(() => {
          console.log("client onReady");
          setTimeout(() => {
              client.onRequest() // your server request
          }, 500);
      });
```

BTW, I understand your troubles very well. I had the same dilemma when I first learned about language server development. Regardless of the vscode-languageclient or vscode-languageserver packages, or the LSP documents, they were all too vague and not conducive to debugging.

**Reply by TracerDS** [reply]
Sorry, but your response didnt answer my issue.
Its not about `onInitialized` event not firing. `onInitialize` (initializ**e**, not initializ**ed**) event doesnt fire.

**Comment by tamuratak** [comment]
> even in [basic LSP example](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-sample) the onInitialize event doesn't fire

The example works for me.

You cannot call `sendNotification` in the `onInitialize` handler because the initialization between the client and the server not finished. You can use `LSPConnection.console.log` instead for a debug purpose.

```ts
LSPConnection.onInitialize((params: lspserver.InitializeParams) => {
    LSPConnection.sendNotification('testnotify', 'initializing...')
})
```

See the spec.

> In addition the server is not allowed to send any requests or notifications to the client until it has responded with an [InitializeResult](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeResult), with the exception that during the initialize request the server is allowed to send the notifications window/showMessage, window/logMessage and [telemetry/event](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#telemetry_event) as well as the window/showMessageRequest request to the client.
>
> https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialize

**Comment by abrilurena** [comment]
Hi @TracerDS! I hope you're doing great. It's been a while since last time, do you still need help with this?

**Reply by TracerDS** [reply]
hmm, it turns out they do fire but the `connection.console.log` just doesnt show in the terminal for some reason.
Its solved ig
